### ISSUE TITLE
Add sparse softmax/log_softmax functionality (ignore zero entries)

### ISSUE THREAD

**Issue by shaibagon** [pytorch/pytorch#23651]
## 🚀 Feature
It would be nice if `torch.sparse` will be extended to include `torch.sparse.min(...)` and `torch.sparse.max(...)` with same interface as existing `torch.sparse.sum(...)`.

For example
```python
s = torch.sparse.FloatTensor(i, v, (height, width))  # make a sparse float tensor
torch.sparse.min(s, dim=1)
# returns minimum along dim=1
torch.sparse.max(s, dim=0)
# returns maximum along dim=0
```

## Motivation
Taking min/max (argmin/argmax if possible) is quite rudimentary. 
For instance suppose I have a sparse matrix `s` with sparse logits and I wish to *robustly* compute the output probabilities using `softmax`. If one wants to avoid overflow/underflow in the exp-sum process, one needs to subtract the `max` across the relevant dimension.
 
## Pitch
Min/max over a selected dimension is quite basic and exists for dense tensor. It would be nice to have similar functionality for sparse tensors as well.

## Alternatives
Very painfully iterate over the entries of the relevant dimension. **Painfully**.

## Additional context
Thinking this over, there is an issue about the "sparse" entries: How should the "sparse" zeros be taken into account? For instance, if all the non-zeros entries in a row are strictly positive, does `min` over this row return 0 (value of "sparse" elements) or the minimal value of the greater than zero elements of the row?

I think, what I am actually looking for is to *ignore* the sparse zeros of the matrix and only consider the values stored in `s._values()` and do `min/max` on these values based on their row/col arrangement in `s`.  
In that case, it is very possible to get `torch.sparse.min(s, dim=0) != torch.min(s.to_dense(), dim=0)`.

cc @ezyang @gchanan @zou3519 @vincentqb

**Comment by CanoeFZH** [comment]
I think it will be good that ```torch.sparse.min(s, dim=..., zero_including=...)``` has parameter with ```zero_including``` to solution the "sparse" zeros issue @shaibagon mentioned.

**Comment by shaibagon** [comment]
Alternatively, it would be nice to have `torch.index_min_`/`torch.index_max_` along the lines of `torch.index_add_`: it's basically the same functionality with slightly less intuitive interface

**Comment by ezyang** [comment]
Is there a reason that `s._values().min()` isn't sufficient for your use case? I guess, it's because you want to take the min over a dimension?

We're not sure what you want to use this for, so more detailed explanation of the use case would help.

**Comment by shaibagon** [comment]
@ezyang `s._values().min()` is indeed insufficient for my use case. I wouldn't have bothered you with such a feature request.

My use case is I have a *sparse* logits matrix with `n` rows and `c` classes/columns (because of the sparse-dense ordering of sparse matrices, the dense batch dimension is last and not first, and I will ignore it in this example).  
I want to compute class probabilities from the *sparse* logits. Since the values of my logits have quite a large range, I cannot simply do `exp` and divide by the row sum - this is extremely unstable numerically (believe me, I tried). The only thing that works is to subtract the `max` of *each row* and the do the `exp` sum.

Thanks.

**Comment by ezyang** [comment]
OK. So, would it be accurate to say, that you would also be satisfied with a sparse softmax? (The change for numerical stability you describe is implemented in our standard softmax kernel.)

**Comment by shaibagon** [comment]
@ezyang a sparse softmax can be awesome.  
However, since I want to *ignore* the sparse zeros `torch.sparse.softmax(X, dim=1) != softmax(X.to_dense(), dim=1)`.

a sparse `log_softmax` will be equally as good. I can also work with `torch.index_max_` version of `torch.index_add_`.

Thank you very much.

**Comment by ezyang** [comment]
OK, I've updated the description here. This feature request seems reasonable; I don't know when the core team will get around to adding it, but we'd be happy to accept a PR that implements it.

**Comment by shaibagon** [comment]
@ezyang
Thank you! 
I'm afraid I only have a very limited implementation of the forward pass w/o gradients.

**Comment by CanoeFZH** [comment]
Seems it's good ramp up to read code of ```torch.sparse.sum(...)``` to get a similar ```torch.sparse.softmax(...)```. Am I right?

**Comment by shaibagon** [comment]
@CanoeFZH I suppose it can. However, I am not very proficient with cuda so I could not benefit from looking at `sparse.sum(...)`'s code. :-(... 
Perhaps someone more proficient than I would be able to do so effectively.

**Comment by accelsao** [comment]
I think we need `coalesce_sparse_cpu/cuda` to support `max/min` reduction, but I cant find suitable function to get the `max/min` value and indices.. Someone can give me some hints?

**Comment by shaibagon** [comment]
I recently came across [pytorch_scatter](https://github.com/rusty1s/pytorch_scatter) package that seems to support these functionalities.

**Comment by ezyang** [comment]
Tentatively bumping priority based on issue activity

**Comment by ezyang** [comment]
This would be transitively solved by #22378

**Comment by pearu** [comment]
Heads up: PR #36305 implements softmax and log_softmax for CPU sparse tensors. Autograd support is included. Early feedback is welcome.

**Comment by pearu** [comment]
Reopening as the PR 36305 resolves the issue for CPU, not for CUDA.
